### PR TITLE
Normalize Categories

### PR DESCRIPTION
**Pull request by sarah-seitanakis**
Add the function that strips descendants from a list of categories or predicates.

### CODE REVIEW

File: reasoner_util/__init__.py [hunk 1]
```diff
@@ -35,3 +38,34 @@ def normalize_ids(curies: List[str]) -> List[str]:
...
+    working_list = item_list.copy()
+    for x_item in item_list:
+        descendants = tk.get_descendants(
+            x_item,
+            reflexive=False,
+            formatted=True,
+        )
+        for desnd in descendants:
+            if desnd in working_list:
+                working_list.remove(desnd)
```
Comment (patrickkwang) [inline reasoner_util/__init__.py:69]: This whole function has worst-case time complexity that's quadratic in the number of items N (`remove()` is linear in the list length), and linear in the average number of descendants D, totaling O(N^2 D). I would have grabbed all of the descendants first (O(ND)) and subsequently filtered the list using a comprehension (O(N)), for a net O(ND). However, your best-case performance is better, and average-case could be up for grabs depending on the typical distribution of items (are there commonly a lot of descendants, and where are they in the list?). This is mostly academic; N and D are upper-bounded in the 10s-100s, so I don't expect speed to be an issue. The other consideration is readability, and I'll let you be the judge there.

File: tests/test_normalizing.py [hunk 2]
```diff
@@ -1,9 +1,41 @@
...
+    catagories = [
+        "biolink:Disease",
+        "biolink:DiseaseOrPhenotypicFeature",
+        "biolink:ThingWithTaxon",
+        "biolink:BiologicalEntity",
+        "biolink:NamedThing",
+        "biolink:Entity",
+      ]
+    output = normalize_qcategories(catagories)
+    assert output == ["biolink:ThingWithTaxon", "biolink:Entity"]
```
Comment (patrickkwang) [inline tests/test_normalizing.py:26]: Note that this assumes that `normalize_qcategories()` does not re-order items.

File: reasoner_util/__init__.py [hunk 3]
```diff
@@ -35,3 +38,34 @@ def normalize_ids(curies: List[str]) -> List[str]:
...
+    return normalized_predicates
+
+
+tk = Toolkit()
+
+
+def strip_descendants(item_list: List[str]) -> List[str]:
+    """strip descendants of biolink catagories or predicates"""
+    working_list = item_list.copy()
+    for x_item in item_list:
```
Comment (patrickkwang) [inline reasoner_util/__init__.py:61]:
```suggestion
    for item in items:
```

File: reasoner_util/__init__.py [hunk 4]
```diff
@@ -1,6 +1,9 @@
-"""Build merging functions."""
+"""Build merging and normalizing functions."""
 from typing import Hashable, Iterable, List, TypeVar
+from typing_extensions import Literal
```
Comment (patrickkwang) [inline reasoner_util/__init__.py:3]: Is this being used? Please remove any unused imports.

**Review by patrickkwang — CHANGES_REQUESTED** [review]

File: reasoner_util/__init__.py (hunk 1 above)
Comment (sarah-seitanakis) [reply reasoner_util/__init__.py]: I just want to make sure I understand--So instead I should get a list of the descendants of all of the items in the input list and then filter out all of the items in the input list that are present in both?

File: tests/test_normalizing.py (hunk 2 above)
Comment (sarah-seitanakis) [reply tests/test_normalizing.py:26]: I didn't think that normalize_qcatagories could re-order items. Should I put in a comment about this assumption or should I write the assertion as output == (["biolink:ThingWithTaxon", "biolink:Entity"] or ["biolink:Entity", "biolink:ThingWithTaxon"]) ?

Comment (patrickkwang) [reply tests/test_normalizing.py:26]: I don't have a preference.

File: reasoner_util/__init__.py (hunk 1 above)
Comment (patrickkwang) [reply reasoner_util/__init__.py]: That's my alternative algorithm in a nutshell, yes.

File: reasoner_util/__init__.py [hunk 5]
```diff
@@ -35,3 +36,34 @@ def normalize_ids(curies: List[str]) -> List[str]:
...
+    for item in items:
+        descendants = tk.get_descendants(
+            item,
+            reflexive=False,
+            formatted=True,
+        )
+        for descendant in descendants:
+            if descendant in working_items:
+                working_items.remove(descendant)
+    stripped_items = working_items
```
Comment (patrickkwang) [inline reasoner_util/__init__.py:68]:
```suggestion
    return working_items
```
?

File: reasoner_util/__init__.py [hunk 6]
```diff
@@ -35,3 +36,36 @@ def normalize_ids(curies: List[str]) -> List[str]:
...
+def strip_descendants(items: List[str]) -> List[str]:
+    """strip descendants of biolink catagories or predicates"""
+    descendants = {
+        descendant
+        for item in items
+        for descendant in tk.get_descendants(
+            item,
+            reflexive=False,
+            formatted=True,
+            )
```
Comment (patrickkwang) [inline reasoner_util/__init__.py:65]:
```suggestion
        )
```

File: reasoner_util/__init__.py [hunk 7]
```diff
@@ -35,3 +36,36 @@ def normalize_ids(curies: List[str]) -> List[str]:
...
+        }
```
Comment (patrickkwang) [inline reasoner_util/__init__.py:66]:
```suggestion
    }
```

**Review by patrickkwang — APPROVED** [review]

**Pull request merged**